sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,687 @@
+
+            higher_sse_index = 1 if lower_sse_index == 0 else 0
+
+            # Centroid with lower SSE is added to result centroids.
+            centroids.append(centers[lower_sse_index])
+
+            if self.verbose:
+                print(f"Centroid Found: {centers[lower_sse_index]}")
+
+            if len(centroids) + 1 == self.n_clusters:
+                # Desired number of cluster centers is reached so centroid
+                # with higher SSE would not be split further
+                centroids.append(centers[higher_sse_index])
+
+                if self.verbose:
+                    print(f"Centroid Found: {centers[higher_sse_index]}")
+                break
+
+            # Data with label that has higher SSE will be further processed
+            indexes_left = (labels == higher_sse_index)
+
+            data_left = data_left[indexes_left]
+            weights_left = weights_left[indexes_left]
+
+        centers = np.asarray(centroids)
+
+        # Restore X
+        if not sp.issparse(X):
+            X += X_mean
+
+        self.n_iter_ = n_iter + 1
+
+        return centers
+
+    def _bisect_by_biggest_sse(self, X, init, random_state, sample_weight):
+        """ Performs Bisecting K-Means, which splits always cluster with
+         biggest SSE from all calculated
+
+            .. note:: All of passed parameters must be pre-calculated
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training instances to cluster.
+
+        init : {'k-means++', 'random'}, callable or ndarray of shape \
+                (n_clusters, n_features)
+                Method for initialization.
+
+        random_state : int, RandomState instance or None, default=None
+            Determines random number generation for centroid initialization.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X.
+
+
+        Returns
+        -------
+        centers : ndarray of shape (n_clusters, n_features)
+            The cluster centers.
+        """
+        # Subtract of mean of X for more accurate distance computations
+        if not sp.issparse(X):
+            X_mean = X.mean(axis=0)
+            X -= X_mean
+
+            if hasattr(init, '__array__'):
+                init -= X_mean
+
+        centers_dict = {
+            0: {'data': X, 'weights': sample_weight, 'sse': None,
+                'centroid': None}
+        }
+
+        init_org = init
+        last_center_id = 0
+
+        for n_iter in range(self.n_clusters - 1):
+
+            biggest, _ = max(centers_dict.items(), key=lambda x: x[1]['sse'])
+
+            # If init array is provided -
+            # Take only part of init that is dedicated for that part of bisect
+            if hasattr(init, '__array__'):
+                init = init_org[n_iter: n_iter + 2].copy()
+                # ^ Not sure if that assigning of init points is
```

Comment:
I don't think it make sense to support arrays in `init`, since presumably the user cannot know what samples are considered at this stage, so they cannot guess what good initial centroids would be.

Reply:
I had doubt about it too.  What do you think, about just replecing it with `UserWarning` that in that case `init` as array won't be supported?

Reply:
I changed the behavior, so only `bisect_strategy = child_biggest_sse` supports `init` param as array. In other cases `ValueError` will be raised